src/k8s/pkg/k8sd/controllers/update_node_configuration.go

```diff
@@ -0,0 +1,107 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/canonical/k8s/pkg/k8sd/types"
+	"github.com/canonical/k8s/pkg/snap"
+	snaputil "github.com/canonical/k8s/pkg/snap/util"
+	"github.com/canonical/k8s/pkg/utils/k8s"
+)
+
+// UpdateNodeConfigurationController asynchronously performs updates of the cluster config.
+// An updates is triggered by sending to the TriggerCh.
+type UpdateNodeConfigurationController struct {
+	snap         snap.Snap
+	waitReady    func()
+	newK8sClient func() (*k8s.Client, error)
+	// TriggerCh is used to trigger config updates.
+	TriggerCh chan struct{}
+	// ReconciledCh is used to indicate that a reconcilation loop has finished.
+	ReconciledCh chan struct{}
+}
+
+// NewUpdateNodeConfigurationController creates a new controller.
+func NewUpdateNodeConfigurationController(snap snap.Snap, waitReady func(), newK8sClient func() (*k8s.Client, error)) *UpdateNodeConfigurationController {
+	return &UpdateNodeConfigurationController{
+		snap:         snap,
+		waitReady:    waitReady,
+		newK8sClient: newK8sClient,
+		TriggerCh:    make(chan struct{}, 1),
+		ReconciledCh: make(chan struct{}, 1),
+	}
+}
+
+func (c *UpdateNodeConfigurationController) retryNewK8sClient(ctx context.Context) (*k8s.Client, error) {
+	for {
+		client, err := c.newK8sClient()
+		if err == nil {
+			return client, nil
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(3 * time.Second):
+		}
+	}
+}
+
+// Run starts the controller.
+// Run accepts a context to manage the lifecycle of the controller.
+// Run accepts a function that retrieves the current cluster configuration.
+// Run will loop everytime the TriggerCh is triggered.
+func (c *UpdateNodeConfigurationController) Run(ctx context.Context, getClusterConfig func(context.Context) (types.ClusterConfig, error)) {
+	c.waitReady()
+
+	client, err := c.retryNewK8sClient(ctx)
+	if err != nil {
+		log.Println(fmt.Errorf("failed to create a Kubernetes client: %w", err))
+	}
+
```

Comment:
move this in the for loop, see https://github.com/canonical/k8s-snap/pull/346/files